crates/db-device/src/device_manager.rs

```diff
@@ -0,0 +1,592 @@
+// Copyright 2020 Alibaba Cloud. All Rights Reserved.
+// Copyright © 2019 Intel Corporation. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! System level device management.
+//!
+//! [IoManager](struct.IoManager.html) is respondsible for managing
+//! all devices of virtual machine, registering IO resources callback,
+//! unregistering devices and helping VM IO exit handling.
+//！
+//！VMM would be responsible for getting device resource request, ask
+//! vm_allocator to allocate the resources, ask vm_device to register the
+//! devices IO ranges, and finally set resources to virtual device.
+
+use std::cmp::{Ord, Ordering, PartialEq, PartialOrd};
+use std::collections::btree_map::BTreeMap;
+use std::fmt::{Display, Formatter};
+use std::ops::Deref;
+use std::result;
+use std::sync::Arc;
+
+use crate::resources::Resource;
+#[cfg(any(target_arch = "x86_64", target_arch = "x86"))]
+use crate::PioAddress;
+use crate::{DeviceIo, IoAddress, IoSize};
+
+/// Error type for `IoManager` usage.
+#[derive(Debug)]
+pub enum Error {
+    /// The inserting device overlaps with a current device.
+    DeviceOverlap,
+    /// The device doesn't exist.
+    NoDevice,
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Error::DeviceOverlap => write!(
+                f,
+                "device_manager: device address conflicts with existing devices"
+            ),
+            Error::NoDevice => write!(f, "device_manager: no such device"),
+        }
+    }
+}
+
+impl std::error::Error for Error {}
+
+/// Simplify the `Result` type.
+pub type Result<T> = result::Result<T, Error>;
+
+/// Structure describing an IO range.
+#[derive(Debug, Copy, Clone, Eq)]
+pub struct IoRange {
+    base: IoAddress,
+    size: IoSize,
+}
+
+impl IoRange {
+    #[cfg(any(target_arch = "x86_64", target_arch = "x86"))]
+    fn new_pio_range(base: u16, size: u16) -> Self {
+        IoRange {
+            base: IoAddress(base as u64),
+            size: IoSize(size as u64),
+        }
+    }
+
+    fn new_mmio_range(base: u64, size: u64) -> Self {
+        IoRange {
+            base: IoAddress(base),
+            size: IoSize(size),
+        }
+    }
+}
+
+impl PartialEq for IoRange {
+    fn eq(&self, other: &IoRange) -> bool {
+        self.base == other.base
+    }
+}
+
+impl Ord for IoRange {
+    fn cmp(&self, other: &IoRange) -> Ordering {
+        self.base.cmp(&other.base)
+    }
+}
+
+impl PartialOrd for IoRange {
+    fn partial_cmp(&self, other: &IoRange) -> Option<Ordering> {
+        self.base.partial_cmp(&other.base)
+    }
+}
+
+/// System IO manager serving for all devices management and VM exit handling.
+#[derive(Clone, Default)]
+pub struct IoManager {
+    #[cfg(any(target_arch = "x86_64", target_arch = "x86"))]
+    /// Range mapping for VM exit pio operations.
+    pio_bus: BTreeMap<IoRange, Arc<dyn DeviceIo>>,
```

Comment:
Could we use the Interval Tree provided by db-allocator to replace the BTreeMap for resource allocation?

Reply:
We need performance test result for this. Interval tree is designed for slow path.

Reply:
got that